RecoCTPPS/PixelLocal/src/RPixPlaneCombinatoryTracking.cc

```diff
@@ -0,0 +1,483 @@
+  possiblePlaneCombinations_.clear();
+}
+
+//------------------------------------------------------------------------------------------------//
+
+void RPixPlaneCombinatoryTracking::initialize(){
+ 
+  uint32_t numberOfCombinations = Factorial(numberOfPlanesPerPot_)/(Factorial(numberOfPlanesPerPot_-trackMinNumberOfPoints_)*Factorial(trackMinNumberOfPoints_));
+  if(verbosity_>=2) std::cout<<"Number of combinations = "<<numberOfCombinations<<std::endl;
+  possiblePlaneCombinations_.reserve(numberOfCombinations);
+
+  possiblePlaneCombinations_ = getPlaneCombinations(listOfAllPlanes_,trackMinNumberOfPoints_);
+
+  if(verbosity_>=2) {
+    for( const auto & vec : possiblePlaneCombinations_){
+      for( const auto & num : vec){
+        std::cout<<num<<" - ";
+      }
+      std::cout<<std::endl;
+    }
+  }
+
+}
+
+//------------------------------------------------------------------------------------------------//
+    
+//This function produces all the possible plane combinations extracting numberToExtract planes over numberOfPlanes planes
+std::vector<std::vector<uint32_t> > RPixPlaneCombinatoryTracking::getPlaneCombinations(std::vector<uint32_t> inputPlaneList, uint32_t numberToExtract)
+{
+    uint32_t numberOfPlanes = inputPlaneList.size();
+    std::string bitmask(numberToExtract, 1); // numberToExtract leading 1's
+    bitmask.resize(numberOfPlanes, 0); // numberOfPlanes-numberToExtract trailing 0's
+    std::vector<std::vector<uint32_t> > planeCombinations;
+
+    // store the combination and permute bitmask
+    do {
+      std::vector<uint32_t> tmpPlaneList;
+      for (uint32_t i = 0; i < numberOfPlanes; ++i) { // [0..numberOfPlanes-1] integers
+        if (bitmask[i]) tmpPlaneList.push_back(inputPlaneList.at(i));
+      }
+        planeCombinations.push_back(tmpPlaneList);
+    } while (std::prev_permutation(bitmask.begin(), bitmask.end()));
+
+    return planeCombinations;
+}
+
+
+//------------------------------------------------------------------------------------------------//
+
+//This function calls it self in order to get all the possible combinations of hits having a certain selected number of planes
+//This function avoids to write for loops in cascade which will not allow to define arbitrarily the minimum number of planes to use
+//The output is stored in a map containing the vector of points and as a key the map of the point forming this vector.
+//This allows to erase the points already used for the track fit
+void RPixPlaneCombinatoryTracking::getHitCombinations(
+  const std::map<CTPPSPixelDetId, std::vector<RPixDetPatternFinder::PointInPlane> > &mapOfAllHits, 
+  std::map<CTPPSPixelDetId, std::vector<RPixDetPatternFinder::PointInPlane> >::iterator mapIterator,
+  std::map<CTPPSPixelDetId, size_t> tmpHitPlaneMap,
+  std::vector<RPixDetPatternFinder::PointInPlane> tmpHitVector,
+  std::map< std::map<CTPPSPixelDetId, size_t>, std::vector<RPixDetPatternFinder::PointInPlane> > &outputMap)
```

Comment:
looks like the type of outputMap should be unordered_map.

What is a typical max permutation count to be sure this would not ever explode?

Reply:
It would be better to linearize/unroll this code instead of recursive calls.
A profiler report may help with an argument that it does not matter much.
Please run it (e.g. igprof with some details in https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Run_profiler_igprof) on events with higher occupancy.
If the data copying is not significant cost for this method, the recursive implementation could stay for now.

Reply:
Dear Slava,
I tried to run igprof but I don't know how to interprete the data.
I copied the results on my lxplus public area:
/afs/cern.ch/user/f/fravera/public/igProfResult.txt
Could you please give a look to it and tell me what is your opinion about this method?
I run it on data collected during the alignment run in which the detector are inserted as close as possible to the beam (half distance with respect to the standard data taking) and so this is the worst case scenario

For which concern the the number of permutations, based on the number we are using now we expect a maximum of 60 hits per pot and minimum 3 planes for a track the maximum number of permutations are 20000. However, this would happen in the extremely unlike case in which all the hits are within the same spatial window of 1 x 1 mm2.
Based on our experience the number of tracks per pattern is extremely rarely higher than 3 which means 3*6 hits (the noise it extremely low) hence the maximum number of permutations is 540. 

The average tracks of physics protons, which are concentrated in a small area of the detector, is below than 2 and events with large number of tracks are due to showeres which are almost uniform on the detector and hence the are divided in separated "roads" before arriving to the track finder.

Reply:
Dear Slava,
I produced the sql:
/afs/cern.ch/user/f/fravera/public/igProfResult.sql3

I tried to sent to you directly the browser link,
but I think I'm missing some steps...

Best regards,
Fabio

Reply:
Hi Slava, as to the first part of your original comment (i.e. the use of unordered_map). I'm wondering if it gives advantages wrt a map. I'm trying to implement it but I find compilation errors of the kind "noexcept(declval<const _Hash&>()(declval<const _Key&>()))>" which makes me suspect I would have to implement hash functions for the "non-standard" objects the map is made of... So, since maps seem to work fine, wouldn't be easier to use them instead of unordered ones? Thanks.